File: Source/Pegasus/Pegasus/Platform/Graphics/OpenGL3/Texture.GL3.cs
namespace Pegasus.Platform.Graphics.OpenGL3
{
	using System;
	using Bindings;
	using Interface;

	/// <summary>
	///     Represents an OpenGL3-based texture stored in GPU memory.
	/// </summary>
	internal unsafe class TextureGL3 : GraphicsObjectGL3, ITexture
	{
		/// <summary>
		///     Initializes a new instance.
		/// </summary>
		/// <param name="graphicsDevice">The graphics device this instance belongs to.</param>
		/// <param name="description">The description of the texture.</param>
		/// <param name="surfaces">The surface data of the texture.</param>
		public TextureGL3(GraphicsDeviceGL3 graphicsDevice, ref TextureDescription description, Surface[] surfaces)
			: base(graphicsDevice)
		{
			Handle = GLContext.Allocate(_gl.GenTextures, "Texture");

			switch (description.Type)
			{
				case TextureType.Texture2D:
				{
					TargetType = GL.Texture_2D;
					_gl.BindTexture(TargetType, Handle);

					if (surfaces != null)
					{
						for (var i = 0; i < description.Mipmaps; ++i)
							UploadTexture(surfaces[i], description.Format, GL.Texture_2D, i);
					}
					else
					{
						var surface = new Surface { Width = description.Width, Height = description.Height };
						UploadTexture(surface, description.Format, GL.Texture_2D, 0);
					}
					break;
				}
				case TextureType.CubeMap:
				{
					TargetType = GL.TextureCubeMap;
					_gl.BindTexture(TargetType, Handle);

					var faces = stackalloc uint[6];
					faces[0] = GL.TextureCubeMapNegativeZ;
					faces[1] = GL.TextureCubeMapNegativeX;
					faces[2] = GL.TextureCubeMapPositiveZ;
					faces[3] = GL.TextureCubeMapPositiveX;
					faces[4] = GL.TextureCubeMapNegativeY;
					faces[5] = GL.TextureCubeMapPositiveY;

					if (surfaces != null)
					{
						for (var i = 0; i < 6; ++i)
						{
							for (var j = 0; j < description.Mipmaps; ++j)
							{
								var index = i * description.Mipmaps + j;
								UploadTexture(surfaces[index], description.Format, faces[i], j);
							}
						}
					}
					else
					{
						var surface = new Surface { Width = description.Width, Height = description.Height };
						for (var i = 0; i < 6; ++i)
							UploadTexture(surface, description.Format, faces[i], 0);
					}
					break;
				}
				default:
					throw new InvalidOperationException("Unsupported texture type.");
			}

			if ((description.Flags & TextureFlags.GenerateMipmaps) == TextureFlags.GenerateMipmaps)
				_gl.GenerateMipmap(GL.Texture_2D);

			RebindTexture();
		}

		/// <summary>
		///     Gets the texture target type.
		/// </summary>
		internal uint TargetType { get; private set; }

		/// <summary>
		///     Gets the texture handle.
		/// </summary>
		internal uint Handle { get; private set; }

		/// <summary>
		///     Binds the texture to the given slot.
		/// </summary>
		/// <param name="slot">The slot the texture should be bound to.</param>
		public void Bind(int slot)
		{
			GraphicsDevice.ChangeActiveTexture(slot);
			_gl.BindTexture(TargetType, Handle);

			GraphicsDevice.State.Texture = Handle;
			GraphicsDevice.State.TextureType = TargetType;
		}

		/// <summary>
		///     Unbinds the texture from the given slot.
		/// </summary>
		/// <param name="slot">The slot the texture should be unbound from.</param>
		public void Unbind(int slot)
		{
			GraphicsDevice.ChangeActiveTexture(slot);
			_gl.BindTexture(TargetType, 0);

			GraphicsDevice.State.Texture = 0;
			GraphicsDevice.State.TextureType = TargetType;
		}

		/// <summary>
		///     Generates the mipmaps for this texture.
		/// </summary>
		public void GenerateMipmaps()
		{
			_gl.BindTexture(TargetType, Handle);
			_gl.GenerateMipmap(TargetType);

			RebindTexture();
		}

		/// <summary>
		///     Sets the debug name of the texture.
		/// </summary>
		/// <param name="name">The debug name of the texture.</param>
		public void SetName(string name)
		{
			// Not supported by OpenGL
		}

		/// <summary>
		///     Disposes the object, releasing all managed and unmanaged resources.
		/// </summary>
		protected override void OnDisposing()
		{
			GLContext.Deallocate(_gl.DeleteTextures, Handle);
		}

		/// <summary>
		///     Uploads a surface to the graphics device.
		/// </summary>
		/// <param name="surface">The surface that should be uploaded.</param>
		/// <param name="format">The format of the surface.</param>
		/// <param name="target">The target texture type.</param>
		/// <param name="level">The resource level the surface should be uploaded to.</param>
		private void UploadTexture(Surface surface, SurfaceFormat format, uint target, int level)
		{
			var internalFormat = format.GetInternalFormat();
			var glFormat = format.Map();
			var type = Texture.IsFloatingPointFormat(format) ? GL.Float : GL.UnsignedByte;

			if (Texture.IsDepthStencilFormat(format))
				type = GL.UnsignedInt248;

			if (Texture.IsCompressedFormat(format))
				_gl.CompressedTexImage2D(target, level, internalFormat, (int)surface.Width, (int)surface.Height, 0, (int)surface.SizeInBytes,
					surface.Data);
			else
				_gl.TexImage2D(target, level, (int)internalFormat, (int)surface.Width, (int)surface.Height, 0, glFormat, type, surface.Data);
		}

		/// <summary>
		///     Rebinds the previously bound texture.
		/// </summary>
		private void RebindTexture()
		{
			if (GraphicsDevice.State.Texture != 0)
				_gl.BindTexture(GraphicsDevice.State.TextureType, GraphicsDevice.State.Texture);
		}
	}
}

File: Source/Pegasus/Platform/Graphics/OpenGL3/TextureGL3.cpp
#include "Prelude.hpp"

using namespace OpenGL3;

Texture* GraphicsDevice::InitializeTexture(TextureDescription* desc, Surface* surfaces)
{
	auto texture = PG_NEW(Texture);
	texture->Obj = Allocate(&GraphicsDevice::glGenTextures, "Texture");
	texture->Format = Map(desc->Format);

	switch (desc->Type)
	{
	case TextureType::Texture2D:
	{
		texture->Type = GL_TEXTURE_2D;
		glBindTexture(texture->Type, texture->Obj);

		if (surfaces != nullptr)
		{
			for (auto i = 0; i < desc->Mipmaps; ++i)
				UploadTexture(surfaces[i], desc->Format, GL_TEXTURE_2D, i);
		}
		else
		{
			Surface surface = { 0 };
			surface.Width = desc->Width;
			surface.Height = desc->Height;

			UploadTexture(surface, desc->Format, GL_TEXTURE_2D, 0);
		}
		break;
	}
	case TextureType::CubeMap:
	{
		texture->Type = GL_TEXTURE_CUBE_MAP;
		glBindTexture(texture->Type, texture->Obj);

		uint32 faces[] =
		{
			GL_TEXTURE_CUBE_MAP_NEGATIVE_Z,
			GL_TEXTURE_CUBE_MAP_NEGATIVE_X,
			GL_TEXTURE_CUBE_MAP_POSITIVE_Z,
			GL_TEXTURE_CUBE_MAP_POSITIVE_X,
			GL_TEXTURE_CUBE_MAP_NEGATIVE_Y,
			GL_TEXTURE_CUBE_MAP_POSITIVE_Y
		};

		if (surfaces != nullptr)
		{
			for (int i = 0; i < 6; ++i)
			{
				for (auto j = 0; j < desc->Mipmaps; ++j)
				{
					auto index = i * desc->Mipmaps + j;
					UploadTexture(surfaces[index], desc->Format, faces[i], j);
				}
			}
		}
		else
		{
			Surface surface = { 0 };
			surface.Width = desc->Width;
			surface.Height = desc->Height;

			for (int i = 0; i < 6; ++i)
				UploadTexture(surface, desc->Format, faces[i], 0);
		}
		break;
	}
	case TextureType::Texture1D:
	case TextureType::Texture3D:
	default:
		PG_NO_SWITCH_DEFAULT;
	}

	if (Enum::HasFlag(desc->Flags, TextureFlags::GenerateMipmaps))
		glGenerateMipmap(GL_TEXTURE_2D);

	RebindTexture();
	return texture;
}

void GraphicsDevice::FreeTexture(Texture* texture)
{
	if (texture == nullptr)
		return;

	glDeleteTextures(1, &texture->Obj);
	PG_DELETE(texture);
}

void GraphicsDevice::BindTexture(Texture* texture, int32 slot)
{
	ChangeActiveTexture(slot);
	glBindTexture(texture->Type, texture->Obj);

	_boundTexture = texture->Obj;
	_boundTextureType = texture->Type;
}

void GraphicsDevice::UnbindTexture(Texture* texture, int32 slot)
{
	ChangeActiveTexture(slot);
	glBindTexture(texture->Type, 0);

	_boundTexture = 0;
	_boundTextureType = texture->Type;
}

void GraphicsDevice::GenerateMipmaps(Texture* texture)
{
	glBindTexture(texture->Type, texture->Obj);
	glGenerateMipmap(texture->Type);

	RebindTexture();
}

void GraphicsDevice::UploadTexture(const Surface& surface, SurfaceFormat format, uint32 target, int32 level)
{
	auto internalFormat = GetInternalFormat(format);
	auto glFormat = Map(format);
	auto type = Graphics::IsFloatingPointFormat(format) ? GL_FLOAT : GL_UNSIGNED_BYTE;

	if (Graphics::IsDepthStencilFormat(format))
		type = GL_UNSIGNED_INT_24_8;

	if (Graphics::IsCompressedFormat(format))
		glCompressedTexImage2D(target, level, internalFormat, surface.Width, surface.Height, 0, surface.SizeInBytes, surface.Data);
	else
	{
		glTexImage2D(
			target, level, safe_static_cast<int32>(internalFormat), surface.Width,
			surface.Height, 0, glFormat, safe_static_cast<uint32>(type), surface.Data);
	}
}

void GraphicsDevice::RebindTexture()
{
	if (_boundTexture != 0)
		glBindTexture(_boundTextureType, _boundTexture);
}

void GraphicsDevice::SetTextureName(Texture* texture, const char* name)
{
	PG_UNUSED(texture);
	PG_UNUSED(name);
}